reporter/src/main/kotlin/utils/FreemarkerTemplateProcessor.kt

```diff
@@ -0,0 +1,242 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.reporter.utils
+
+import freemarker.cache.ClassTemplateLoader
+import freemarker.template.Configuration
+import freemarker.template.TemplateExceptionHandler
+
+import java.io.File
+
+import kotlin.reflect.full.memberProperties
+
+import org.ossreviewtoolkit.model.Identifier
+import org.ossreviewtoolkit.model.OrtResult
+import org.ossreviewtoolkit.model.licenses.LicenseConfiguration
+import org.ossreviewtoolkit.model.licenses.LicenseView
+import org.ossreviewtoolkit.model.licenses.ResolvedLicense
+import org.ossreviewtoolkit.model.licenses.ResolvedLicenseFileInfo
+import org.ossreviewtoolkit.model.licenses.ResolvedLicenseInfo
+import org.ossreviewtoolkit.model.licenses.filterExcluded
+import org.ossreviewtoolkit.reporter.ReporterInput
+import org.ossreviewtoolkit.utils.expandTilde
+import org.ossreviewtoolkit.utils.log
+
+class FreemarkerTemplateProcessor(
+    private val filePrefix: String,
+    private val defaultTemplateName: String
+) {
+    companion object {
+        private const val OPTION_TEMPLATE_ID = "template.id"
+        private const val OPTION_TEMPLATE_PATH = "template.path"
+    }
+
+    fun generateOutput(input: ReporterInput, outputDir: File, options: Map<String, String>): List<File> {
+        val projects = input.ortResult.getProjects().map { project ->
+            PackageModel(project.id, input)
+        }
+
+        val packages = input.ortResult.getPackages().map { pkg ->
+            PackageModel(pkg.pkg.id, input)
+        }
+
+        val dataModel = mapOf(
+            "projects" to projects,
+            "packages" to packages,
+            "ortResult" to input.ortResult,
+            "licenseTextProvider" to input.licenseTextProvider,
+            "helper" to TemplateHelper(input.ortResult, input.licenseConfiguration)
+        )
+
+        val freemarkerConfig = Configuration(Configuration.VERSION_2_3_30).apply {
+            defaultEncoding = "UTF-8"
+            fallbackOnNullLoopVariable = false
+            logTemplateExceptions = true
+            tagSyntax = Configuration.SQUARE_BRACKET_TAG_SYNTAX
+            templateExceptionHandler = TemplateExceptionHandler.RETHROW_HANDLER
+            templateLoader = ClassTemplateLoader(
+                this@FreemarkerTemplateProcessor.javaClass.classLoader,
+                "template"
```

Comment:
Just a reminder, if using "templates" (plural, see below) then this also needs to be reflected here.